testutils/src/main/java/com/microsoft/identity/internal/testutils/shadows/ShadowHttpClient.java

```diff
@@ -0,0 +1,115 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.internal.testutils.shadows;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.microsoft.identity.common.internal.net.AbstractHttpClient;
+import com.microsoft.identity.common.internal.net.HttpClient;
+import com.microsoft.identity.common.internal.net.HttpResponse;
+import com.microsoft.identity.common.internal.net.UrlConnectionHttpClient;
+import com.microsoft.identity.internal.testutils.HttpRequestInterceptor;
+import com.microsoft.identity.internal.testutils.MockHttpClient;
+
+import org.robolectric.annotation.Implementation;
+import org.robolectric.annotation.Implements;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Map;
+
+/**
+ * Allows us to mock http request responses by shadowing the {@link HttpClient}.
+ *
+ * @see MockHttpClient - for setting up interceptors and mock responses
+ * @see HttpRequestInterceptor - an implementation for intercepting http requests.
+ */
+@Implements(AbstractHttpClient.class)
+public class ShadowHttpClient {
+
+    public HttpResponse method(@NonNull HttpClient.HttpMethod httpMethod,
+                               @NonNull URL requestUrl,
+                               @NonNull Map<String, String> requestHeaders,
+                               @Nullable byte[] requestContent) throws IOException {
+        HttpRequestInterceptor interceptor = MockHttpClient.intercept(httpMethod, requestUrl);
+        if (interceptor == null) {
+            return UrlConnectionHttpClient.getDefaultInstance().method(httpMethod, requestUrl, requestHeaders, requestContent);
+        } else {
+            return interceptor.intercept(httpMethod, requestUrl, requestHeaders, requestContent);
+        }
+    }
+
+    @Implementation
+    public HttpResponse put(@NonNull URL requestUrl,
+                            @NonNull Map<String, String> requestHeaders,
+                            @Nullable byte[] requestContent) throws IOException {
+        return method(HttpClient.HttpMethod.PUT, requestUrl, requestHeaders, requestContent);
+    }
```

Comment:
Does any of these need to be written?

Reply:
(As in, can we not rely on the superclass implementations?)

Reply:
The Shadow has to re-implement the methods in order for the mock to be used. Otherwise, not having them, the superclass will make the actual Http request.